### ISSUE TITLE
Out-of-date CL headers in CL/ can confuse compilers

### ISSUE THREAD

**Issue by jasonsewall-intel** [UoB-HPC/BabelStream#109]
The headers in CL are quite out of date, and I've found that they cause issues when compiling with a modern toolchain (eg. intel/llvm).  

I would recommend that this be removed completely or replaced with a submodule that is kept close to upstream.

**Comment by tom91136** [comment]
The CL headers were added for both ComputeCpp support and the OCL implementation. 
That said, I think something isn't quite right with dpcpp here.
What happened was that dpcpp prepends the *source file's base directory* as part of the include path.
If I do `dpcpp -v -DSYCL -std=c++17 -o foo.o -c /home/tom/babelstream-upstream/SYCLStream.cpp`, the search path looks like this:
```

...
"/opt/intel/oneapi/compiler/2021.4.0/linux/bin/clang++" -cc1  .... -I /home/tom/babelstream-upstream ....
...

#include "..." search starts here:
#include <...> search starts here:
 /home/tom/babelstream-upstream
 /opt/intel/oneapi/vpl/2021.6.0/include
 /opt/intel/oneapi/tbb/2021.4.0/env/../include
 /opt/intel/oneapi/mpi/2021.4.0//include
 /opt/intel/oneapi/mkl/2021.4.0/include
 /opt/intel/oneapi/ipp/2021.4.0/include
 /opt/intel/oneapi/ippcp/2021.4.0/include
 /opt/intel/oneapi/dpl/2021.5.0/linux/include
 /opt/intel/oneapi/dpcpp-ct/2021.4.0/include
 /opt/intel/oneapi/dnnl/2021.4.0/cpu_dpcpp_gpu_dpcpp/lib
 /opt/intel/oneapi/dev-utilities/2021.4.0/include
 /opt/intel/oneapi/dal/2021.4.0/include
 /opt/intel/oneapi/ccl/2021.4.0/include/cpu_gpu_dpcpp
 /opt/intel/oneapi/compiler/2021.4.0/linux/bin/../include/sycl
 /opt/intel/oneapi/compiler/2021.4.0/linux/include
 /opt/intel/oneapi/compiler/2021.4.0/linux/bin/../compiler/include
 /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11
 /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/x86_64-redhat-linux
 /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/backward
 /opt/intel/oneapi/compiler/2021.4.0/linux/lib/clang/13.0.0/include
 /usr/local/include
 /usr/include

```
The `CL` directory happens to be at that path and the include guard prevented dpcpp's `/include/sycl/CL/sycl/detail/cl.h` from loading the correct header.

---
Here's a reduced version of the issue, consider:
```
bad_include
├── bad.cpp
└── stdio.h
```

```c++
// bad.cpp
#include <stdio.h>
int main(){
    return 42;
}
```

```c
// stdio.h
#pragma once
#error should not be included
```
And then:
```bash
> cd bad_include

> clang++ bad.cpp # ok

> g++ bad.cpp # ok

> source  /opt/intel/oneapi/setvars.sh

> dpcpp bad.cpp
In file included from bad.cpp:2:
/home/tom/Desktop/bad_include/stdio.h:2:2: error: should not be included
#error should not be included
 ^
1 error generated. 


> clang++ bad.cpp -I . # if we include the header manually, then this fails as expected:
In file included from bad.cpp:2:
./stdio.h:2:2: error: should not be included
#error should not be included


> g++ bad.cpp -I . # if we include the header manually, then this fails as expected:
In file included from bad.cpp:2:
./stdio.h:2:2: error: #error should not be included
    2 | #error should not be included
      |  ^~~~~

> clang++ --version  
clang version 12.0.1 (Fedora 12.0.1-1.fc34)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

> g++ --version                                              
g++ (GCC) 11.2.1 20210728 (Red Hat 11.2.1-1)
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

> dpcpp --version  
Intel(R) oneAPI DPC++/C++ Compiler 2021.4.0 (2021.4.0.20210924)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/intel/oneapi/compiler/2021.4.0/linux/bin

```
This happens on both Fedora 34 and Ubuntu 20.04 so it's probably not configuration related.

**Comment by tom91136** [comment]
Ok, I think the root cause was https://github.com/intel/llvm/issues/4226, which broke Kokkos as well

**Comment by tom91136** [comment]
@tomdeakin this only affects intel/llvm (dpcpp) so I think we can workaround this by putting the CL headers a level deeper (e.g `includes/CL/...` so that it doesn't get included first.

**Comment by tom91136** [comment]
PS: CI didn't catch this because it was pulling a commit of intel/llvm before https://github.com/intel/llvm/pull/4071

**Comment by tomdeakin** [comment]
@jasonsewall-intel  See above for a potential DPC++ issue.

@tom91136 What happens if we delete the `CL/` folder entirely, and require the headers are provided by the user? One issue might be `cl2.hpp`, which doesn't come as standard, but we could add only that one (as a submodule?) and not the other OpenCL headers? What do you think?

**Comment by jasonsewall-intel** [comment]
I don't work on the compiler, but it does seem like the issue was fomented by https://github.com/intel/llvm/pull/4071/commits/f764e543a20197093a9da8cb349ce7784de781d0, which explicitly tries to resolve includes in the CWD before looking elsewhere. 

C++ says that searches are done in "implementation-defined places", which isn't very helpful IMHO but there are probably good reasons for doing so. I do think that it makes sense that a user may want to prioritize resolving conflicts by putting an include in the source directory, and that that is a reasonable interpretation of what Babelstream does right now. 

I think a good way to resolve it would be to just keep the local CL/ directory up-to-date, at least with some regularity. Alternately, you could place it in something like includes/CL and then specific -I`pwd`/includes or something like that when a given system needs to be handed CL headers.

**Comment by tomdeakin** [comment]
We've resolved this in #102.
The headers are downloaded freshly at build time *iff* they are required and not already installed as part of a specific combination of model/platform.
E.g. For DPC++, it should *not* download them and use the ones that come along with the DPC++ installation.
